src/komodo_structs.cpp

```diff
@@ -0,0 +1,348 @@
+        throw parse_error("Unable to read nid of event U from file");
+    if (fread(&mask, 1, sizeof(mask), fp) != sizeof(mask))
+        throw parse_error("Unable to read mask of event U from file");
+    if (fread(&hash, 1, sizeof(hash), fp) != sizeof(hash))
+        throw parse_error("Unable to read hash of event U from file");
+}
+
+std::ostream& operator<<(std::ostream& os, const event_u& in)
+{
+    const event& e = dynamic_cast<const event&>(in);
+    os << e;
+    os << in.n << in.nid;
+    os.write((const char*)in.mask, 8);
+    os.write((const char*)in.hash, 32);
+    return os;
+}
+
+event_kmdheight::event_kmdheight(uint8_t* data, long &pos, long data_len, int32_t height, bool includeTimestamp) : event(EVENT_KMDHEIGHT, height)
+{
+    mem_read(this->kheight, data, pos, data_len);
+    if (includeTimestamp)
+        mem_read(this->timestamp, data, pos, data_len);
+}
+
+event_kmdheight::event_kmdheight(FILE *fp, int32_t height, bool includeTimestamp) : event(EVENT_KMDHEIGHT, height)
+{
+    if ( fread(&kheight,1,sizeof(kheight),fp) != sizeof(kheight) )
+        throw parse_error("Unable to parse KMD height");
+    if (includeTimestamp)
+    {
+        if ( fread( &timestamp, 1, sizeof(timestamp), fp) != sizeof(timestamp) )
+            throw parse_error("Unable to parse timestamp of KMD height");
+    }
+}
+
+std::ostream& operator<<(std::ostream& os, const event_kmdheight& in)
+{
+    const event& e = dynamic_cast<const event&>(in);
+    os << e << serializable<int32_t>(in.kheight);
+    if (in.timestamp > 0)
+        os << serializable<int32_t>(in.timestamp);
+    return os;
+}
+
+event_opreturn::event_opreturn(uint8_t *data, long &pos, long data_len, int32_t height) : event(EVENT_OPRETURN, height)
+{
+    mem_read(this->txid, data, pos, data_len);
+    mem_read(this->vout, data, pos, data_len);
+    mem_read(this->value, data, pos, data_len);
+    uint16_t oplen;
+    mem_read(oplen, data, pos, data_len);
+    if (oplen <= data_len - pos)
+    {
+        this->opret = std::vector<uint8_t>( &data[pos], &data[pos] + oplen);
+        pos += oplen;
+    }
+}
+
+event_opreturn::event_opreturn(FILE* fp, int32_t height) : event(EVENT_OPRETURN, height)
+{
+    if ( fread(&txid,1,sizeof(txid),fp) != sizeof(txid) )
+        throw parse_error("Unable to parse txid of opreturn record");
+    if ( fread(&vout,1,sizeof(vout),fp) != sizeof(vout) )
+        throw parse_error("Unable to parse vout of opreturn record");
+    if ( fread(&value,1,sizeof(value),fp) != sizeof(value) )
+        throw parse_error("Unable to parse value of opreturn record");
+    uint16_t oplen;
+    if ( fread(&oplen,1,sizeof(oplen),fp) != sizeof(oplen) )
+        throw parse_error("Unable to parse length of opreturn record");
+    std::unique_ptr<uint8_t> b(new uint8_t[oplen]);
+    if ( fread(b.get(), 1, oplen, fp) != oplen)
+        throw parse_error("Unable to parse binary data of opreturn");
+    this->opret = std::vector<uint8_t>( b.get(), b.get() + oplen);
```

Comment:
Is any reason why we need dynamic array allocation here, using std::unique_ptr and read `opret` content into intermediate variable `b`? I mean, as we have `oplen` known we could do something like direct reading in vector:

```
        uint16_t oplen;
        if ( fread(&oplen,1,sizeof(oplen),fp) != sizeof(oplen) )
            throw parse_error("Unable to parse length of opreturn record");
        this->opret.resize(oplen);
        if ( fread(&this->opret[0], sizeof this->opret[0], this->opret.size(), fp) != oplen )
            throw parse_error("Unable to parse binary data of opreturn");
```

Reply:
Good solution. Thanks.